RTN-023.tex

```diff
@@ -50,6 +50,372 @@
+	a DP0.2 processing would split the 40,000 visits into 1000 visit
+	groups.  Consideration should be given to how long (wallclock time)
+	a single campaign 'unit' may be run and to not generally exceed 
+	24 hours to a few days to avoid interruptions from downtimes.
+	Also depends on ability of system to recover from interruptions and
+	continue without needing to restart the whole campaign from
+	the beginning.
+	
+
+	\item  Should allow(?) 'last minute editing' of configs before
+	submission to tweak something, change a pipeline version or
+	visit input list
+
+	\item each campaign should be identified by some unique id string, 
+	perhaps same as workflow manager identifying string?
+
+	\end{itemize}
+
+	\item Ability to store and manage a set of BPS submit scripts 
+	(or equivalent).
+	For instance, as each BPS submit yaml is performed, the yaml
+	script contents could be saved in a database with metadata such
+	as time of submission, location of physical processing
+	(SLAC, IN2P3, UK, Cloud, user-laptop), current state (non started,
+	running, finished-no-errors, finished-with-errors)
+
+
+	\item Ability to quickly access and view logs of each 
+	campaign's individual tasks/jobs to determine errors
+
+	\item Ability to view overall state of running and recently
+	completed campaigns, without being overwhelmed by long-ago completed
+	or abandoned campaigns.
+
+	\item As runs may be distributed 'world-wide' for execution,
+	local sites should be able to view only those jobs running at
+	their local site on local resources or get a wider 'world-wide' view.
+
+	\item Ability to straight-forwardly dump dataids (or filenames
+	with paths) and metadata (exposureid, filter) of inputs or outputs
+	from some specific job or task of the processing campaign to
+	track down issues and errors.  This ability may be present already
+	but some 'helper' scripts could simplify the process.
+
+	\item Ability to watch, in realtime, as jobs or tasks within
+	a campaign or set of campaigns are running.  There is a tool	
+	called 'bps report' which does this at some level, showing how
+	many individual inputs in a campaign set have been processed
+	so far.
+
+	\end {enumerate}
+
+\section{BPS and the PanDA system}
+
+For DP0 \cite{RTN-013}, Rubin is using 
+the \cite{PanDA} workload data.  The Batch Production 
+Service (BPS) \cite{LDM-636} is used as an interface to PanDA.  
+
+BPS:
+\begin{enumerate}
+
+\item Accepts a yaml formatted keyword-value set of parameters which
+defines a single run. (see example run specification yaml below).
+
+\item After 'bps submit <run-spec.yaml>', queries the butler to generate
+a 'quantum graph', a DAG which defines the ordered set of 
+pipeline/pipetask which are to be run on which specific set of input 
+collections (or subsets of a collection which may be picked out via SQL
+type where clauses)
+
+\item Generates an ordered list of tasks, each of which may consist
```

Comment:
This use of "task' is confusing given that the jobs that are created run `PipelineTask` which run `Task`.  Isn't this what we are calling "job"?

Reply:
bps creates a Generic Workflow - a graph containing where each node is a job (as defined by @timj) and edges indicate job dependencies. The nodes have information such as command lines, requested resources (e.g., memory, cpus). This Generic Workflow is passed to the WMS plugins (e.g., HTCondor/DAGman, PanDA) to be "converted" to their own workflow format to be submitted for execution. The WMS need to know nothing about what's inside the jobs (e.g., which Quanta are being executed).

Reply:
this is helpful definition stuff -- I will try and incorporate -- eventually.